soc/arm/st_stm32/stm32g4/power.c

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2021 STMicroelectronics.
+ * Copyright (c) 2021 G-Technologies Sdn. Bhd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/pm.h>
+#include <soc.h>
+
+#include <clock_control/clock_stm32_ll_common.h>
+#include <stm32g4xx_ll_bus.h>
+#include <stm32g4xx_ll_cortex.h>
+#include <stm32g4xx_ll_pwr.h>
+#include <stm32g4xx_ll_system.h>
+#include <stm32g4xx_ll_utils.h>
+LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
+
+/* Invoke Low Power/System Off specific Tasks */
+__weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
+	if (state != PM_STATE_SUSPEND_TO_IDLE) {
+		LOG_DBG("Unsupported power state %u", state);
+		return;
+	}
+
+	switch (substate_id) {
+	case 1: /* this corresponds to the STOP0 mode: */
+		/* enter STOP0 mode */
+		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP0);
+		LL_LPM_EnableDeepSleep();
+		/* enter SLEEP mode : WFE or WFI */
+		k_cpu_idle();
+		break;
+	case 2: /* this corresponds to the STOP1 mode: */
+		/* enter STOP1 mode */
+		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP1);
+		LL_LPM_EnableDeepSleep();
+		/* enter SLEEP mode : WFE or WFI */
+		k_cpu_idle();
+		break;
+	default:
+		LOG_DBG("Unsupported power state substate-id %u", substate_id);
+		break;
+	}
+}
+
+/* Handle SOC specific activity after Low Power Mode Exit */
+__weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+{
+	if (state != PM_STATE_SUSPEND_TO_IDLE) {
+		LOG_DBG("Unsupported power substate %u", state);
+	} else {
+		switch (substate_id) {
+		case 1: /* STOP0 */
+			__fallthrough;
+		case 2: /* STOP1 */
+			LL_LPM_DisableSleepOnExit();
+			/* Clear SLEEPDEEP bit */
+			LL_LPM_EnableSleep();
+			break;
+		default:
+			LOG_DBG("Unsupported power substate-id %u", substate_id);
+			break;
+		}
+		/* need to restore the clock */
+		stm32_clock_control_init(NULL);
+	}
+
+	/*
+	 * System is now in active mode.
+	 * Reenable interrupts which were disabled
+	 * when OS started idling code.
+	 */
+	irq_unlock(0);
+}
+
+/* Initialize STM32 Power */
+static int stm32_power_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* enable Power clock */
+	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
+
+	/* keep in mind that debugging draws a lot of power */
+#ifdef CONFIG_DEBUG
+	LL_DBGMCU_EnableDBGStopMode();
+#else
+	LL_DBGMCU_DisableDBGStopMode();
```

Comment:
Disable is usually not performed in that case, but I ack this is probably a good thing, as it will disable bit systematically at boot and avoid surprises that could be seen when using openocd (which sets this bit during flash) or flashing a debug disabled SW back to back with a debug enabled SW.
We probably need to extend this to other series (after testing though, as this DBGMCU IP block implementation is full of surprises).

Reply:
I am not sure what you want me to do here?

Reply:
Nothing. This work for me to apply this on other series.